### ISSUE TITLE
Canceling paid payment succeed without error

### ISSUE THREAD

**Issue by fmarchalemisys** [VivaPayments/API#2360]
Hi,

I'm calling `DELETE https://demo.vivapayments.com/api/orders/8096464623190719` ([documented here](https://developer.vivawallet.com/apis-for-payments/payment-api/#tag/Payments/paths/~1api~1orders~1{orderCode}/delete)) to cancel a payment.

I called that endpoint on a **paid** payment to see what would happen.

The payment isn't canceled. That's good.

But the above endpoint returns a success.

```
{"OrderCode":8096464623190719,"ErrorCode":0,"ErrorText":null,"TimeStamp":"2023-04-01T16:52:18.9695779+03:00","CorrelationId":null,"EventId":0,"Success":true}
```

Shouldn't it return a failure?

**Comment by VivaWalletCustomerSupport** [comment]
Hello @fmarchalemisys,

Thank you for your message. Are you making the API request with the merchant's demo account credentials or with ISV demo account credentials?

Kind regards,
Viva Wallet E-commerce Support Team

**Comment by fmarchalemisys** [comment]
Hi @VivaWalletCustomerSupport ,

The payment was started with an ISV demo account but I was trying to cancel it with the merchant demo account.

I quickly patched my code to cancel it using the [ISV api call](https://developer.vivawallet.com/isv-partner-program/payment-isv-api/#tag/Payments/paths/~1api~1orders~1{orderCode}/delete) but the response is `{"Message":"Authorization has been denied for this request."}`. Both on a paid or unpaid payment.

The basic authentication token is computed from `base64("isv_client_id:merchant_id:isv_secret")`. If I'm correct, it's apparently the only thing I have to change to cancel with the ISV account instead of the merchant account.

I called the endpoint to cancel a paid payment with

```
curl -X DELETE 'https://demo.vivapayments.com/api/orders/7912667195117930' -H 'authorization: Basic dDV…`
```

And to cancel a pending payment:

```
curl -X DELETE 'https://demo.vivapayments.com/api/orders/5312771018629118' -H 'authorization: Basic dDV…'
```

**Comment by fmarchalemisys** [comment]
I notice my ISV client ID (demo account) looks like `letters_and_digits.apps.vivapayments.com` when the documentation says the reseller ID should looks like `121212-abcd-1212-abcd-121212121212`.

So, I guess I'm not using the correct ID. It must be that extra set of credentials you wrote about in another of my issue ticket :unamused: .

Unless you can spot an obvious mistake I'm making, I'll get in touch with my representative about that reseller ID.

**Comment by VivaWalletCustomerSupport** [comment]
Hello @fmarchalemisys,

Thank you for the additional information. If you wish to cancel the transaction via [Cancel payment order for ISV](https://developer.vivawallet.com/isv-partner-program/payment-isv-api/#tag/Payments/paths/~1api~1orders~1{orderCode}/delete), the credentials you should use are provided by us. Please contact your Sales representative or send us an e-mail to support@vivawallet.com with the Github case number (#2360) on the subject of the e-mail.

Kind regards,
Viva Wallet E-commerce Support Team

**Comment by fmarchalemisys** [comment]
In fact, my initial question is not ISV related. Let me rephrase it :slightly_smiling_face: 

I can cancel a **pending** payment using the merchant's basic authentication token despite the payment being created by the ISV account.

The problem I'm writing about is that the API returns a success when cancelling a **paid** payment that can't be canceled. I'm expecting some error message instead of a success.

Or should I expect the paid payment to be canceled instead?

**Comment by VivaWalletCustomerSupport** [comment]
Hello @fmarchalemisys,

Thank you for the message. Kindly note that merchants can cancel a transaction that initiated by an ISV account. The result of this API request should be the cancellation of the transaction and not an error message. 

Kind regards,
Viva Wallet E-commerce Support Team

**Comment by fmarchalemisys** [comment]
Hi @VivaWalletCustomerSupport ,

Thanks for the clarification.

As far as I can see in the merchant's web interface a paid payment is not canceled by the api call despite de success returned by the api call.

In other words, the api says "the payment was canceled" but the web interface shows it isn't.

I expect a paid payment cannot be canceled. Hence my question about the apparent success returned by `DELETE https://demo.vivapayments.com/api/orders/`.

**Comment by VivaWalletCustomerSupport** [comment]
Hello @fmarchalemisys,

Thank you for the prompt response. As we can observe, the transaction with order code 8096464623190719 is cancelled. Did you cancel it manually via your Viva Wallet account? If you didn't cancel it manually then it was cancelled via the API call as intended.

Kind regards,
Viva Wallet E-commerce Support Team

**Comment by fmarchalemisys** [comment]
Hi @VivaWalletCustomerSupport ,

I probably tried to cancel the payment from the demo account's web interface to see if the payment was cancelable :thinking: 

To be on the safe side, I just started a new payment with order code 5122919810716733. It is paid and I leave it as such.

When I run this curl command:

```
curl -X DELETE 'https://demo.vivapayments.com/api/orders/5122919810716733' -H 'Authorization: Basic NWY…'
```

the answer is:

```
{"OrderCode":5122919810716733,"ErrorCode":0,"ErrorText":null,"TimeStamp":"2023-04-04T19:19:11.6186871+03:00","CorrelationId":null,"EventId":0,"Success":true}
```

So, the DELETE api call returns a success but the payment isn't canceled.

**Comment by VivaWalletCustomerSupport** [comment]
Hello @fmarchalemisys,

Thank you for your message. Further to the previous conversation (https://github.com/VivaPayments/API/issues/2361) our [Cancel payment order API request](https://developer.vivawallet.com/apis-for-payments/payment-api/#tag/Payments/paths/~1api~1orders~1{orderCode}/delete) cancels a standing payment order ("StateId" with value 0 (Pending)) and not a paid order code. If you wish to cancel a specific transaction kindly use our [Cancel transaction API request](https://developer.vivawallet.com/apis-for-payments/payment-api/#tag/Transactions/paths/~1api~1transactions~1{transaction_id}/delete).

Kind regards,
Viva Wallet E-commerce Support Team

**Comment by fmarchalemisys** [comment]
Hi @VivaWalletCustomerSupport 

Sorry for being thick, but I don't get it :slightly_frowning_face: 

I don't want to cancel the payment. I just want to get informed that the cancellation failed because the payment has been paid.

My purpose is to avoid this race condition:

1. I check the status of the payment.
2. The api says it is pending.
3. I call the api to cancel the payment.
4. Customer completes the payment somewhere between 2 and 3 so the payment can't actually be canceled.
5. I receive a cancel failure form the api so I stop canceling the payment on my side and, instead, start processing the customer's purchase.

I guess I can add step 6 to call the api and double check the payment status after the supposed successful cancellation but that's a waste of bandwidth and it takes one more roundtrip time.

Hence my question. Is the cancel success a bug in the api or am I missing something?

**Comment by VivaWalletCustomerSupport** [comment]
Hello @fmarchalemisys,

Thank you for the additional information. We are investigating your request. We will inform you as soon as we have any updates.

Kind regards,
Viva Wallet E-commerce Support Team

**Comment by VivaWalletCustomerSupport** [comment]
Hello @fmarchalemisys,

Thank you for waiting. The response you are getting is correct since the call is correct. If the payment order status is paid ("StateId" with value 3 (paid)) you will get the same response, but the order code won't be canceled. 
You can use the [retrieve order](https://developer.vivawallet.com/apis-for-payments/payment-api/#tag/Payments/paths/~1api~1orders~1{orderCode}/get) to verify that.

Kind regards,
Viva Wallet E-commerce Support Team

**Comment by fmarchalemisys** [comment]
Hi @VivaWalletCustomerSupport ,

Thank you for confirming a success is the correct response when the api call does not complete the task.

I'll look for another solution.
